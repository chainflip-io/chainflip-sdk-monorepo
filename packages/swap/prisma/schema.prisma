// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
