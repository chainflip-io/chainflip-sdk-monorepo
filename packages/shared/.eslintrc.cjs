// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require('path');

module.exports = {
  extends: '../../.eslintrc.json',
  plugins: ['eslint-plugin-n'],
  rules: {
    'n/no-process-env': ['error'],
    'import/no-extraneous-dependencies': [
      'error',
      {
        packageDir: [__dirname, path.join(__dirname, '..', '..')],
      },
    ],
  },
  overrides: [
    {
      files: ['*.test.ts', '*.mjs', '*.js', '*.cjs'],
      rules: {
        'n/no-process-env': ['off'],
      },
    },
  ],
};
