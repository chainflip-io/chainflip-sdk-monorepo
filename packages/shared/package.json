{
  "name": "@chainflip-io/sdk-shared",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "rm -rf dist",
    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.mjs.json",
    "build:abis": "typechain --target=ethers-v5 --out-dir=./src/abis ./abis/**",
    "lint": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "test": "jest",
    "test:ci": "CI=1 pnpm test -- --forceExit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "zod": "^3.22.2"
  },
  "dependencies": {
    "@polkadot/util": "^12.4.2",
    "@polkadot/util-crypto": "^12.4.2",
    "ethers": "5.7.2",
    "rxjs": "^7.8.1",
    "ws": "^8.13.0"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@typechain/ethers-v5": "^11.1.1",
    "@types/ws": "^8.5.5",
    "typechain": "^8.3.1",
    "winston": "^3.10.0",
    "zod": "^3.22.2"
  }
}
