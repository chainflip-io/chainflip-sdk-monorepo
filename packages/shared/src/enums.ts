type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const InternalAssets = arrayToMap(['Flip', 'Usdc', 'Dot', 'Eth', 'Btc']);
export type InternalAsset =
  (typeof InternalAssets)[keyof typeof InternalAssets];

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'backspin',
  'sisyphos',
  'perseverance',
  'mainnet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetConstants = {
  [InternalAssets.Eth]: {
    chain: Chains.Ethereum,
    asset: Assets.ETH,
    name: 'Ether',
    decimals: 18,
    contractId: 1,
  },
  [InternalAssets.Flip]: {
    chain: Chains.Ethereum,
    asset: Assets.FLIP,
    name: 'FLIP',
    decimals: 18,
    contractId: 2,
  },
  [InternalAssets.Usdc]: {
    chain: Chains.Ethereum,
    asset: Assets.USDC,
    name: 'USDC',
    decimals: 6,
    contractId: 3,
  },
  [InternalAssets.Dot]: {
    chain: Chains.Polkadot,
    asset: Assets.DOT,
    name: 'Polkadot',
    decimals: 10,
    contractId: 4,
  },
  [InternalAssets.Btc]: {
    chain: Chains.Bitcoin,
    asset: Assets.BTC,
    name: 'Bitcoin',
    decimals: 8,
    contractId: 5,
  },
} satisfies Record<
  InternalAsset,
  {
    chain: Chain;
    asset: Asset;
    name: string;
    decimals: number;
    contractId: number;
  }
>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainConstants = {
  [Chains.Ethereum]: {
    assets: [Assets.ETH, Assets.FLIP, Assets.USDC],
    gasAsset: Assets.ETH,
    contractId: 1,
  },
  [Chains.Polkadot]: {
    assets: [Assets.DOT],
    gasAsset: Assets.DOT,
    contractId: 2,
  },
  [Chains.Bitcoin]: {
    assets: [Assets.BTC],
    gasAsset: Assets.BTC,
    contractId: 3,
  },
} satisfies Record<
  Chain,
  { assets: Asset[]; gasAsset: Asset; contractId: number }
>;

export type AssetOfChain<C extends Chain> =
  (typeof chainConstants)[C]['assets'][number];

export type UncheckedAssetAndChain = { asset: Asset; chain: Chain };
export type AssetAndChain = {
  [C in Chain]: { chain: C; asset: AssetOfChain<C> };
}[Chain];

export type ChainAssetMap<T> = {
  [C in Chain]: {
    [A in AssetOfChain<C>]: T;
  };
};

export function isValidAssetAndChain(
  assetAndChain: UncheckedAssetAndChain,
): assetAndChain is AssetAndChain {
  const { asset, chain } = assetAndChain;
  const validAssets = chainConstants[chain].assets as string[];

  return validAssets.includes(asset);
}

export function assertIsValidAssetAndChain(
  assetAndChain: UncheckedAssetAndChain,
): asserts assetAndChain is AssetAndChain {
  if (!isValidAssetAndChain(assetAndChain)) {
    throw new Error('invalid asset and chain combination');
  }
}

export const readChainAssetMap = <T>(
  value: ChainAssetMap<T>,
  asset: UncheckedAssetAndChain,
): T => {
  assertIsValidAssetAndChain(asset);
  const chainValues = value[asset.chain];
  return chainValues[asset.asset as keyof typeof chainValues];
};

export function getInternalAsset(asset: UncheckedAssetAndChain) {
  const map: ChainAssetMap<InternalAsset> = {
    [Chains.Ethereum]: {
      [Assets.USDC]: InternalAssets.Usdc,
      [Assets.FLIP]: InternalAssets.Flip,
      [Assets.ETH]: InternalAssets.Eth,
    },
    [Chains.Bitcoin]: {
      [Assets.BTC]: InternalAssets.Btc,
    },
    [Chains.Polkadot]: {
      [Assets.DOT]: InternalAssets.Dot,
    },
  };

  return readChainAssetMap(map, asset);
}
