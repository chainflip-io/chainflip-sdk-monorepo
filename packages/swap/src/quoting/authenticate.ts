import * as crypto from 'crypto';
import type { Server } from 'socket.io';
import { promisify } from 'util';
import { z } from 'zod';
import { InternalAssetMap, getInternalAsset, InternalAsset, InternalAssets } from '@/shared/enums';
import { isNotNullish } from '@/shared/guards';
import { assetAndChain } from '@/shared/parsers';
import prisma from '../client';
import { type QuotingSocket } from './Quoter';
import baseLogger from '../utils/logger';

const logger = baseLogger.child({ module: 'quoter' });

const verifyAsync = promisify(crypto.verify);

type Middleware = Parameters<Server['use']>[0];
type Next = Parameters<Middleware>[1];

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

const mapAssets = (quotedAssets: InternalAsset[] | null): InternalAssetMap<boolean> => {
  assert(quotedAssets === null || quotedAssets.length !== 0, 'no assets quoted');

  return Object.fromEntries(
    Object.values(InternalAssets).map(
      (asset) => [asset, quotedAssets === null || quotedAssets.includes(asset)] as const,
    ),
  ) as InternalAssetMap<boolean>;
};

const authSchema = z.union([
  z
    .object({
      client_version: z.literal('1'),
      market_maker_id: z.string(),
      timestamp: z.number(),
      signature: z.string(),
    })
    .transform((data) => ({
      ...data,
      quoted_assets: mapAssets(null),
      account_id: data.market_maker_id,
    })),
  z.object({
    client_version: z.literal('2'),
    account_id: z.string(),
    timestamp: z.number(),
    signature: z.string(),
    quoted_assets: z
      .array(assetAndChain.transform(getInternalAsset))
      .transform((assets) => mapAssets(assets.filter(isNotNullish))),
  }),
]);

const parseKey = (key: string) => {
  try {
    return crypto.createPublicKey({ key: Buffer.from(key), format: 'pem', type: 'spki' });
  } catch {
    throw new Error('invalid public key');
  }
};

const authenticate = async (socket: QuotingSocket, next: Next) => {
  try {
    const result = authSchema.safeParse(socket.handshake.auth);

    logger.info('received auth', { auth: result.data, error: result.error?.message });

    assert(result.success, 'invalid auth');

    const auth = result.data;
    const timeElapsed = Date.now() - auth.timestamp;

    logger.info('time elapsed from handshake timestamp', { timeElapsed });

    assert(timeElapsed < 30_000 && timeElapsed >= -30_000, 'invalid timestamp');

    const marketMaker = await prisma.marketMaker.findUnique({
      where: { name: auth.account_id },
    });

    assert(marketMaker, 'market maker not found');

    const key = parseKey(marketMaker.publicKey);

    const signaturesMatch = await verifyAsync(
      null,
      Buffer.from(`${auth.account_id}${auth.timestamp}`, 'utf8'),
      key,
      Buffer.from(auth.signature, 'base64'),
    );

    assert(signaturesMatch, 'invalid signature');

    // https://socket.io/docs/v4/server-socket-instance/#socketdata
    // eslint-disable-next-line no-param-reassign
    socket.data = {
      marketMaker: marketMaker.name,
      quotedAssets: auth.quoted_assets,
      clientVersion: auth.client_version,
    };

    next();
  } catch (error) {
    logger.warn('authentication error', { error: (error as Error).message });
    next(error as Error);
  }
};

export default authenticate;
