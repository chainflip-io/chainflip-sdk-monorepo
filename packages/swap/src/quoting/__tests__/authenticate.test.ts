import * as crypto from 'crypto';
import { promisify } from 'util';
import { vi, describe, it, beforeEach, expect, Mock } from 'vitest';
import { InternalAssetMap, InternalAssets } from '@/shared/enums';
import prisma from '../../client';
import authenticate from '../authenticate';

const generateKeyPairAsync = promisify(crypto.generateKeyPair);

describe(authenticate, () => {
  let next: Mock;
  let privateKey: crypto.KeyObject;

  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE private."MarketMaker" CASCADE`;
    next = vi.fn();
    const pair = await generateKeyPairAsync('ed25519');
    await prisma.marketMaker.create({
      data: {
        name: 'web_team_whales',
        publicKey: pair.publicKey.export({ format: 'pem', type: 'spki' }).toString('base64'),
      },
    });
    privateKey = pair.privateKey;
  });

  it.each([
    {},
    { client_version: '1' },
    { market_maker_id: 'web_team_whales' },
    { timestamp: Date.now() },
    { signature: 'deadbeef' },
    { client_version: '1', market_maker_id: 'web_team_whales' },
    { client_version: '1', timestamp: Date.now() },
    { client_version: '1', signature: 'deadbeef' },
    { market_maker_id: 'web_team_whales', timestamp: Date.now() },
    { market_maker_id: 'web_team_whales', signature: 'deadbeef' },
    { timestamp: Date.now(), signature: 'deadbeef' },
    {
      client_version: '1',
      market_maker_id: 'web_team_whales',
      timestamp: Date.now(),
    },
  ])('rejects invalid auth shape', async (auth) => {
    await authenticate({ handshake: { auth } } as any, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(new Error('invalid auth'));
  });

  it.each([[-30001, 30000]])('rejects invalid timestamps', async (diff) => {
    await authenticate(
      {
        handshake: {
          auth: {
            client_version: '1',
            market_maker_id: 'web_team_whales',
            timestamp: Date.now() + diff,
            signature: 'deadbeef',
          },
        },
      } as any,
      next,
    );
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(new Error('invalid timestamp'));
  });

  it('rejects unknown market maker', async () => {
    await authenticate(
      {
        handshake: {
          auth: {
            client_version: '1',
            market_maker_id: 'unknown',
            timestamp: Date.now(),
            signature: 'deadbeef',
          },
        },
      } as any,
      next,
    );
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(new Error('market maker not found'));
  });

  it('rejects invalid public key', async () => {
    await prisma.marketMaker.update({
      where: { name: 'web_team_whales' },
      data: { publicKey: 'invalid' },
    });

    await authenticate(
      {
        handshake: {
          auth: {
            client_version: '1',
            market_maker_id: 'web_team_whales',
            timestamp: Date.now(),
            // test different lengths
            signature: 'deadbeef',
          },
        },
      } as any,
      next,
    );

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(new Error('invalid public key'));
  });

  it.each([
    {
      client_version: '1',
      market_maker_id: 'web_team_whales',
      timestamp: Date.now(),
      // test different lengths
      signature: 'deadbeef',
    },
    {
      client_version: '1',
      market_maker_id: 'web_team_whales',
      timestamp: Date.now(),
      // test same lengths
      signature: 'deadbeefdeadbeefdeadbeefdeadbeef',
    },
  ])('rejects invalid signature', async (auth) => {
    await authenticate(
      {
        handshake: {
          auth,
        },
      } as any,
      next,
    );
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(new Error('invalid signature'));
  });

  it('accepts valid v1 authentication', async () => {
    const timestamp = Date.now();
    const name = 'web_team_whales';

    const signature = crypto
      .sign(null, Buffer.from(`${name}${timestamp}`, 'utf8'), privateKey)
      .toString('base64');

    const socket = {
      handshake: {
        auth: {
          client_version: '1',
          market_maker_id: name,
          timestamp,
          signature,
        },
      },
    };

    await authenticate(socket as any, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith();
    expect((socket as any).data).toStrictEqual({
      marketMaker: name,
      quotedAssets: Object.fromEntries(Object.values(InternalAssets).map((asset) => [asset, true])),
      clientVersion: '1',
    });
  });

  it('accepts valid v2 authentication', async () => {
    const timestamp = Date.now();
    const name = 'web_team_whales';

    const signature = crypto
      .sign(null, Buffer.from(`${name}${timestamp}`, 'utf8'), privateKey)
      .toString('base64');

    const socket = {
      handshake: {
        auth: {
          client_version: '2',
          account_id: name,
          timestamp,
          signature,
          quoted_assets: [{ chain: 'Ethereum', asset: 'FLIP' }],
        },
      },
    };

    const quotedAssets = Object.fromEntries(
      Object.values(InternalAssets).map((asset) => [asset, false]),
    ) as InternalAssetMap<boolean>;
    quotedAssets.Flip = true;

    await authenticate(socket as any, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith();
    expect((socket as any).data).toStrictEqual({
      marketMaker: name,
      quotedAssets,
      clientVersion: '2',
    });
  });
});
