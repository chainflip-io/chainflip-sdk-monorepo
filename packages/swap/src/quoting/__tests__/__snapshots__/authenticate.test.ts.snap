// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`authenticate > accepts valid v2 authentication ({ beta: false, useMevFactor: true, mevFactors: [ { asset: 'Flip', factor: 1, side: 'SELL' }, { asset: 'Flip', factor: -1, side: 'BUY' } ] }) 1`] = `
{
  "beta": false,
  "clientVersion": "2",
  "marketMaker": "web_team_whales",
  "mevFactors": {
    "buy": {
      "Flip": -1,
    },
    "sell": {
      "Flip": -1,
    },
  },
  "quotedAssets": {
    "ArbEth": false,
    "ArbUsdc": false,
    "Btc": false,
    "Dot": false,
    "Eth": false,
    "Flip": true,
    "HubDot": false,
    "HubUsdc": false,
    "HubUsdt": false,
    "Sol": false,
    "SolUsdc": false,
    "Usdc": false,
    "Usdt": false,
  },
}
`;

exports[`authenticate > accepts valid v2 authentication ({ beta: false, useMevFactor: true, mevFactors: [ { asset: 'Flip', factor: 1, side: 'SELL', type: 'MEV' }, { asset: 'Flip', factor: -1, side: 'BUY', type: 'MEV' } ] }) 1`] = `
{
  "beta": false,
  "clientVersion": "2",
  "marketMaker": "web_team_whales",
  "mevFactors": {
    "buy": {
      "ArbEth": 0,
      "ArbUsdc": 0,
      "Btc": 0,
      "Dot": 0,
      "Eth": 0,
      "Flip": -1,
      "HubDot": 0,
      "HubUsdc": 0,
      "HubUsdt": 0,
      "Sol": 0,
      "SolUsdc": 0,
      "Usdc": 0,
      "Usdt": 0,
    },
    "sell": {
      "ArbEth": 0,
      "ArbUsdc": 0,
      "Btc": 0,
      "Dot": 0,
      "Eth": 0,
      "Flip": -1,
      "HubDot": 0,
      "HubUsdc": 0,
      "HubUsdt": 0,
      "Sol": 0,
      "SolUsdc": 0,
      "Usdc": 0,
      "Usdt": 0,
    },
  },
  "quotedAssets": {
    "ArbEth": false,
    "ArbUsdc": false,
    "Btc": false,
    "Dot": false,
    "Eth": false,
    "Flip": true,
    "HubDot": false,
    "HubUsdc": false,
    "HubUsdt": false,
    "Sol": false,
    "SolUsdc": false,
    "Usdc": false,
    "Usdt": false,
  },
  "replenishmentFactors": {},
}
`;

exports[`authenticate > accepts valid v2 authentication ({ beta: true, useMevFactor: false, mevFactors: [ { asset: 'Flip', factor: 1, side: 'SELL' }, { asset: 'Flip', factor: -1, side: 'BUY' } ] }) 1`] = `
{
  "beta": true,
  "clientVersion": "2",
  "marketMaker": "web_team_whales",
  "mevFactors": {
    "buy": {},
    "sell": {},
  },
  "quotedAssets": {
    "ArbEth": false,
    "ArbUsdc": false,
    "Btc": false,
    "Dot": false,
    "Eth": false,
    "Flip": true,
    "HubDot": false,
    "HubUsdc": false,
    "HubUsdt": false,
    "Sol": false,
    "SolUsdc": false,
    "Usdc": false,
    "Usdt": false,
  },
}
`;

exports[`authenticate > accepts valid v2 authentication ({ beta: true, useMevFactor: false, mevFactors: [ { asset: 'Flip', factor: 1, side: 'SELL', type: 'MEV' }, { asset: 'Flip', factor: -1, side: 'BUY', type: 'MEV' } ] }) 1`] = `
{
  "beta": true,
  "clientVersion": "2",
  "marketMaker": "web_team_whales",
  "mevFactors": {
    "buy": {
      "ArbEth": 0,
      "ArbUsdc": 0,
      "Btc": 0,
      "Dot": 0,
      "Eth": 0,
      "Flip": 0,
      "HubDot": 0,
      "HubUsdc": 0,
      "HubUsdt": 0,
      "Sol": 0,
      "SolUsdc": 0,
      "Usdc": 0,
      "Usdt": 0,
    },
    "sell": {
      "ArbEth": 0,
      "ArbUsdc": 0,
      "Btc": 0,
      "Dot": 0,
      "Eth": 0,
      "Flip": 0,
      "HubDot": 0,
      "HubUsdc": 0,
      "HubUsdt": 0,
      "Sol": 0,
      "SolUsdc": 0,
      "Usdc": 0,
      "Usdt": 0,
    },
  },
  "quotedAssets": {
    "ArbEth": false,
    "ArbUsdc": false,
    "Btc": false,
    "Dot": false,
    "Eth": false,
    "Flip": true,
    "HubDot": false,
    "HubUsdc": false,
    "HubUsdt": false,
    "Sol": false,
    "SolUsdc": false,
    "Usdc": false,
    "Usdt": false,
  },
  "replenishmentFactors": {},
}
`;
